CSS style 'filter: drop-shadow...' does not render correctly without repaint.

vitorroriz opened the pull request:
#### 9ee234dfdeeb2ab91b7c0e8124d7f981e2039918
<pre>
CSS style &apos;filter: drop-shadow...&apos; does not render correctly without repaint.
<a href="https://bugs.webkit.org/show_bug.cgi?id=207586">https://bugs.webkit.org/show_bug.cgi?id=207586</a>
rdar://59430460

Reviewed by NOBODY (OOPS!).

Fix by having filter outsets contribute to visual overflow. This first patch is the simplest
case: adding to visual overflow on the element with the filter. Future patches will handle
the case where filter on some ancestor needs to contribute to visual overflow on
descendants.&quot;

The addition here is regarding &quot;handle the case where filter on some ancestor needs to contribute
to visual overflow on descendants&quot;.

We are using a internal property named &quot;accumulatedFilterOutsets&quot; to accumulate the maximum
outsets between the element&apos;s parent outsets and the element&apos;s own outsets, which is updated during
style building (StyleBuilderCustom). That way, we don&apos;t need to traverse the render tree to calculate
the maximum outsets along the path between a node and the tree&apos;s root.

At RenderBox::applyVisualEffectOverflow we will take the accumulated filters outsets in consideration
when calculating the element&apos;s visual overflow.

* LayoutTests/compositing/filters/sw-layer-overlaps-hw-shadow-expected.txt:
* LayoutTests/compositing/filters/sw-nested-shadow-overlaps-hw-nested-shadow-expected.txt:
* LayoutTests/compositing/filters/sw-shadow-overlaps-hw-layer-expected.txt:
* LayoutTests/compositing/filters/sw-shadow-overlaps-hw-shadow-expected.txt:
* LayoutTests/fast/repaint/drop-shadow-box-change-expected.txt: Added.
* LayoutTests/fast/repaint/drop-shadow-box-change.html: Added.
* LayoutTests/fast/repaint/drop-shadow-change-expected.txt: Added.
* LayoutTests/fast/repaint/drop-shadow-change.html: Added.
* LayoutTests/fast/repaint/drop-shadow-writing-modes-expected.txt: Added.
* LayoutTests/fast/repaint/drop-shadow-writing-modes.html: Added.
* LayoutTests/platform/mac-wk1/fast/repaint/drop-shadow-box-change-expected.txt: Added.
* LayoutTests/platform/mac-wk1/fast/repaint/drop-shadow-change-expected.txt: Added.
* LayoutTests/platform/mac-wk1/fast/repaint/drop-shadow-with-descendants-change-expected.txt: Added.
* LayoutTests/platform/mac-wk1/fast/repaint/drop-shadow-writing-modes-expected.txt: Added.
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/platform/graphics/filters/FilterOperations.cpp:
(WebCore::FilterOperations::outsets const):
(WebCore::FilterOperations::accumulatedFilterOutsets const):
* Source/WebCore/platform/graphics/filters/FilterOperations.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::addVisualEffectOverflow):
(WebCore::RenderBox::applyVisualEffectOverflow const):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::accumulatedFilterOutsets const):
(WebCore::RenderStyle::hasAccumulatedFilterOutsets const):
(WebCore::RenderStyle::setAccumulatedFilterOutsets):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueFilter):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ee234dfdeeb2ab91b7c0e8124d7f981e2039918

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3616 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3657 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3806 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5042 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3911 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3590 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3757 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3705 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3140 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3660 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3901 "2 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3262 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4865 "Built successfully") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1407 "2 failures") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3234 "4 failures") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/3300 "142 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3208 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3293 "3 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4652 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3669 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2971 "1 flakes 3 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3203 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3234 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3235 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3482 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/platform/graphics/filters/FilterOperations.cpp:
```diff
@@ -110,6 +110,21 @@ IntOutsets FilterOperations::outsets() const
+    auto left = std::max(outsets.left(), parentOutsets.left());
+    auto bottom = std::max(outsets.bottom(), parentOutsets.bottom());
+    return { top, right, bottom, left };
+}
```
Is this the correct way to combine outsets for vertical or flipped blocks writing modes?

vitorroriz committed:
CSS style 'filter: drop-shadow...' does not render correctly without repaint.
https://bugs.webkit.org/show_bug.cgi?id=207586
rdar://59430460

Reviewed by NOBODY (OOPS!).

Fix by having filter outsets contribute to visual overflow. This first patch is the simplest
case: adding to visual overflow on the element with the filter. Future patches will handle
the case where filter on some ancestor needs to contribute to visual overflow on
descendants."

The addition here is regarding "handle the case where filter on some ancestor needs to contribute
to visual overflow on descendants".

We are using a internal property named "accumulatedFilterOutsets" to accumulate the maximum
outsets between the element's parent outsets and the element's own outsets, which is updated during
style building (StyleBuilderCustom). That way, we don't need to traverse the render tree to calculate
the maximum outsets along the path between a node and the tree's root.

At RenderBox::applyVisualEffectOverflow we will take the accumulated filters outsets in consideration
when calculating the element's visual overflow.

* LayoutTests/compositing/filters/sw-layer-overlaps-hw-shadow-expected.txt:
* LayoutTests/compositing/filters/sw-nested-shadow-overlaps-hw-nested-shadow-expected.txt:
* LayoutTests/compositing/filters/sw-shadow-overlaps-hw-layer-expected.txt:
* LayoutTests/compositing/filters/sw-shadow-overlaps-hw-shadow-expected.txt:
* LayoutTests/fast/repaint/drop-shadow-box-change-expected.txt: Added.
* LayoutTests/fast/repaint/drop-shadow-box-change.html: Added.
* LayoutTests/fast/repaint/drop-shadow-change-expected.txt: Added.
* LayoutTests/fast/repaint/drop-shadow-change.html: Added.
* LayoutTests/fast/repaint/drop-shadow-writing-modes-expected.txt: Added.
* LayoutTests/fast/repaint/drop-shadow-writing-modes.html: Added.
* LayoutTests/platform/mac-wk1/fast/repaint/drop-shadow-box-change-expected.txt: Added.
* LayoutTests/platform/mac-wk1/fast/repaint/drop-shadow-change-expected.txt: Added.
* LayoutTests/platform/mac-wk1/fast/repaint/drop-shadow-with-descendants-change-expected.txt: Added.
* LayoutTests/platform/mac-wk1/fast/repaint/drop-shadow-writing-modes-expected.txt: Added.
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/platform/graphics/filters/FilterOperations.cpp:
(WebCore::FilterOperations::outsets const):
(WebCore::FilterOperations::accumulatedFilterOutsets const):
* Source/WebCore/platform/graphics/filters/FilterOperations.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::addVisualEffectOverflow):
(WebCore::RenderBox::applyVisualEffectOverflow const):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::accumulatedFilterOutsets const):
(WebCore::RenderStyle::hasAccumulatedFilterOutsets const):
(WebCore::RenderStyle::setAccumulatedFilterOutsets):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueFilter):
